lib/tests/test_annotate.rs

```diff
@@ -0,0 +1,248 @@
+
+    // 4    "2 1 3"
+    // |\
+    // | 3  "1 3"
+    // | |
+    // 2 |  "2 1"
+    // |/
+    // 1    "1"
+    let mut tx = repo.start_transaction(&settings);
+    let mut create_commit = create_commit_fn(tx.repo_mut(), &settings);
+    let content1 = "1\n";
+    let content2 = "2\n1\n";
+    let content3 = "1\n3\n";
+    let content4 = "2\n1\n3\n";
+    let tree1 = create_tree(repo, &[(file_path, content1)]);
+    let tree2 = create_tree(repo, &[(file_path, content2)]);
+    let tree3 = create_tree(repo, &[(file_path, content3)]);
+    let tree4 = create_tree(repo, &[(file_path, content4)]);
+    let commit1 = create_commit("commit1", &[root_commit_id], tree1.id());
+    let commit2 = create_commit("commit2", &[commit1.id()], tree2.id());
+    let commit3 = create_commit("commit3", &[commit1.id()], tree3.id());
+    let commit4 = create_commit("commit4", &[commit2.id(), commit3.id()], tree4.id());
+    drop(create_commit);
+
+    insta::assert_snapshot!(annotate(tx.repo(), &commit4, file_path), @r#"
+    commit2: 2
+    commit1: 1
+    commit3: 3
+    "#);
+}
+
+#[test]
+fn test_anntoate_merge_split() {
+    let settings = testutils::user_settings();
+    let test_repo = TestRepo::init();
+    let repo = &test_repo.repo;
+
+    let root_commit_id = repo.store().root_commit_id();
+    let file_path = RepoPath::from_internal_string("file");
+
+    // 4    "2 1a 1b 3 4"
+    // |\
+    // | 3  "1b 3"
+    // | |
+    // 2 |  "2 1a"
+    // |/
+    // 1    "1a 1b"
+    let mut tx = repo.start_transaction(&settings);
+    let mut create_commit = create_commit_fn(tx.repo_mut(), &settings);
+    let content1 = "1a\n1b\n";
+    let content2 = "2\n1a\n";
+    let content3 = "1b\n3\n";
+    let content4 = "2\n1a\n1b\n3\n4\n";
+    let tree1 = create_tree(repo, &[(file_path, content1)]);
+    let tree2 = create_tree(repo, &[(file_path, content2)]);
+    let tree3 = create_tree(repo, &[(file_path, content3)]);
+    let tree4 = create_tree(repo, &[(file_path, content4)]);
+    let commit1 = create_commit("commit1", &[root_commit_id], tree1.id());
+    let commit2 = create_commit("commit2", &[commit1.id()], tree2.id());
+    let commit3 = create_commit("commit3", &[commit1.id()], tree3.id());
+    let commit4 = create_commit("commit4", &[commit2.id(), commit3.id()], tree4.id());
+    drop(create_commit);
+
+    insta::assert_snapshot!(annotate(tx.repo(), &commit4, file_path), @r#"
+    commit2: 2
+    commit1: 1a
+    commit1: 1b
+    commit3: 3
+    commit4: 4
+    "#);
+}
+
+#[test]
+#[should_panic] // FIXME: shouldn't panic because of duplicated "1"s
+fn test_anntoate_merge_dup() {
+    let settings = testutils::user_settings();
+    let test_repo = TestRepo::init();
+    let repo = &test_repo.repo;
+
+    let root_commit_id = repo.store().root_commit_id();
+    let file_path = RepoPath::from_internal_string("file");
+
+    // 4    "2 1 1 3 4"
+    // |\
+    // | 3  "1 3"
+    // | |
+    // 2 |  "2 1"
+    // |/
+    // 1    "1"
```

Comment:
Added this test and `test_annotate_merge_simple()`. I'll fix the panic in later patch.